src/env.rs

```diff
@@ -0,0 +1,218 @@
+    All,
+    /// Fully encapsulate the environment, removing all host
+    /// environment import checks
+    None,
+    /// Only allow the provided environment variable keys
+    Allow(Vec<String>),
+    /// Allow all environment variables, except the provided keys
+    Deny(Vec<String>),
+}
+
+impl WasiVirt {
+    fn get_or_create_env(&mut self) -> &mut VirtEnv {
+        self.virt_opts.env.get_or_insert_with(Default::default)
+    }
+
+    pub fn env_host_allow(mut self, allow_list: &[&str]) -> Self {
+        let env = self.get_or_create_env();
+        env.host = HostEnv::Allow(allow_list.iter().map(|s| s.to_string()).collect());
+        self
+    }
+
+    pub fn env_host_deny(mut self, deny_list: &[&str]) -> Self {
+        let env = self.get_or_create_env();
+        env.host = HostEnv::Deny(deny_list.iter().map(|s| s.to_string()).collect());
+        self
+    }
+
+    pub fn env_host_all(mut self) -> Self {
+        let env = self.get_or_create_env();
+        env.host = HostEnv::All;
+        self
+    }
+
+    pub fn env_host_none(mut self) -> Self {
+        let env = self.get_or_create_env();
+        env.host = HostEnv::None;
+        self
+    }
+
+    pub fn env_overrides(mut self, overrides: &[(&str, &str)]) -> Self {
+        let env = self.get_or_create_env();
+        for (key, val) in overrides {
+            env.overrides.push((key.to_string(), val.to_string()));
+        }
+        self
+    }
+}
+
+pub fn create_env_virt<'a>(module: &'a mut Module, env: &VirtEnv) -> Result<()> {
+    let env_ptr_addr = {
+        let env_ptr_export = module
+            .exports
+            .iter()
+            .find(|expt| expt.name.as_str() == "env")
+            .context("Adapter 'env' is not exported")?;
+        let ExportItem::Global(env_ptr_global) = env_ptr_export.item else {
+            bail!("Adapter 'env' not a global");
+        };
+        let GlobalKind::Local(InitExpr::Value(Value::I32(env_ptr_addr))) =
+            &module.globals.get(env_ptr_global).kind
+        else {
+            bail!("Adapter 'env' not a local I32 global value");
+        };
+        *env_ptr_addr as u32
+    };
+
+    // If host env is disabled, remove its import entirely
+    // replacing it with a stub panic
+    if matches!(env.host, HostEnv::None) {
+        stub_imported_func(module, "wasi:cli-base/environment", "get-environment")?;
+        // we do arguments as well because virt assumes reactors for now...
+        stub_imported_func(module, "wasi:cli-base/environment", "get-arguments")?;
+    }
+
+    let memory = module
+        .memories
+        .iter()
+        .nth(0)
+        .context("Adapter does not export a memory")?
+        .id();
+
+    // prepare the field data list vector for writing
+    // strings must be sorted as binary searches are used against this data
+    let mut field_data_vec: Vec<&str> = Vec::new();
+    let mut sorted_overrides = env.overrides.clone();
+    sorted_overrides.sort_by_key(|(k, _)| k.to_string());
```

Comment:
Oops, I made a mistake here in my previous suggestion. To avoid the allocation:

```suggestion
    sorted_overrides.sort_by(|(a, _), (b, _)| a.cmp(b));
```
```

Reply:
Thanks, less worried about allocations in this part of the code though as well!